ddht/v5_1/alexandria/client.py

```diff
@@ -431,6 +461,25 @@ async def get_content(
         )
         return response
 
+    async def advertise(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        *,
+        advertisements: Collection[Advertisement],
+    ) -> Tuple[AckMessage, ...]:
+        advertisement_batches = partition_advertisements(
+            advertisements, max_payload_size=MAX_PAYLOAD_SIZE,
+        )
+        messages = tuple(AdvertiseMessage(batch) for batch in advertisement_batches)
+        assert len(messages)
+        return tuple(
+            [
+                await self._request(node_id, endpoint, message, AckMessage)
```

Comment:
context gathering: `ACK` is only used to respond to advertisements? obviously a request for content is acknowledged by returning that content, but why is it important to ack the ad?

Reply:
I'm not sure that the ack is required.  

The original reason that I implement the `Ack` message was to communicate the recipient's updated `advertisement_radius` back to the sender of the advertisement.  The sender uses the `advertisement_radius` to determine whether a given advertisement is *interesting* to a node.  Supposing that I sent you an advertisement that caused something else to get evicted from your advertisement database, your advertisement radius would shrink, and subsequent ads might no longer be interesting to you.  In a perfect would, this is an effective mechanism to prevent someone sending a bunch of advertisements to a node that the node doesn't care about due to a full advertisement db and the ads being less proximate than their current ones.

But... it's looking a lot like validation of an advertisement might take longer than is reasonable to wait for the ack, meaning that the receiving node will need to ack the advertisement before actually validating it, or they sending node will timeout the request/response waiting for the ack due to validation taking too long on the recipient's end.

Since timing out waiting for the Ack isn't ideal, I've got things currently implemented such that the ack is sent prior to full validation...  Which is still better than nothing since it can still give the sending node an updated view of my advertisment radius, but... once the advertisment validates the radius might shrink and the sending node won't know that until they either send another advertisment **or** do a ping/pong.